Implement for Windows

wfraser opened the pull request:
Windows has the `_wexecvp` function in the C runtime which is the same as Unix's `execvp` but takes Windows wide strings for arguments instead.

The biggest difference is that instead of using `CString` to make FFI-ready strings, we use `std::os::windows::ffi::OsStrExt::encode_wide` which builds a `Vec<u16>` for us, but we need to check for interior nulls manually, and can't return the same `NulError` that comes from `CString`. So I've changed the error type slightly on Windows accordingly.

This isn't a breaking change because this crate didn't build on Windows before anyway :)

Unix functionality is unaffected.

As a slight drive-by improvement, I fixed up the `std::error::Error` method implementations to remove the deprecated `description()` implementation and use `source()` instead of the deprecated `cause()`.

wfraser committed:
implement for Windows

using _wexecvp() in Windows' CRT

wfraser committed:
better error type for Windows

wfraser commented:
re: #3

emk commented on src/lib.rs:
```diff
@@ -29,44 +30,49 @@ use std::os::unix::ffi::OsStrExt;
 pub enum Error {
     /// One of the strings passed to `execv` contained an internal null byte
     /// and can't be passed correctly to C.
-    BadArgument(NulError),
+    #[cfg(unix)] BadArgument(NulError),
+    #[cfg(windows)] BadArgument,
```
It might be awkward to have incompatible error types on different platforms. It would be better to use `BadArgument` with no argument, and bump the crate version to acknowledge a breaking change, then it would be to have different error signatures on different platforms.

If we have more than one type of `BadArgument`, we might also split this into multiple values:

```
enum Error {
    NullByteInArgument,
    SomeOtherBadArgument,
    Errno(Errno),
}
```

wfraser replied:
If you're okay with breaking changes, then yes I agree. The `NulError` tells you where the nul is, but in practice that isn't really helpful, and can be left out.

emk commented on src/lib.rs:
```diff
@@ -29,44 +30,49 @@ use std::os::unix::ffi::OsStrExt;
-        }
-    }
-    fn cause(&self) -> Option<&error::Error> {
+    #[cfg(unix)]
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
```
`source` should be the same on Unix and Windows.

wfraser replied:
With the change to `Error` it will be removed entirely as there will be no interior error to forward to any more :)

emk commented on src/lib.rs:
```diff
@@ -29,44 +30,49 @@ use std::os::unix::ffi::OsStrExt;
     fn from(err: NulError) -> Error {
         Error::BadArgument(err)
     }
 }
```
This could probably be removed, and replaced with `map_err` in the appropriate places.

emk commented on src/lib.rs:
```diff
@@ -123,6 +139,50 @@ pub fn execvp<S, I>(program: S, args: I) -> Error
+    let mut arg_ptrs: Vec<_> = args_wide.iter().map(|arg| arg.as_ptr()).collect();
+    arg_ptrs.push(ptr::null());
+
+    extern "C" {
+        // https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/execvp-wexecvp
+        fn _wexecvp(cmdname: *const libc::wchar_t, argv: *const *const libc::wchar_t)
+            -> libc::intptr_t;
+    }
```
Is there a Windows `-sys` library that provides this definition already?

If there isn't, we should probably at least move it to the top level of this file.

wfraser replied:
Ideally it would be in the `libc` crate, but it isn't. Agree with moving it up to top-level.

wfraser replied:
I've got a PR open to add these to libc now: https://github.com/rust-lang/libc/pull/1796

emk reviewed (changes requested):
Interesting! Thank you for working on this patch.

This is a fairly old library, and since Rust 1.9.0, `exec` has been [built into Rust](https://doc.rust-lang.org/std/os/unix/process/trait.CommandExt.html#tymethod.exec), but only for Unix-like systems. So this library had been largely deprecated. (I somehow missed updating the `Cargo.toml` file to indicate this.)

I think there are a few important questions here:

1. The [docs say that](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/execvp-wexecvp?view=vs-2019) "This API cannot be used in applications that execute in the Windows Runtime. For more information, see CRT functions not supported in Universal Windows Platform apps." Are normal Rust Windows apps supposed to work with the Windows Runtime? Are there any special `cfg` flags that we need to check or document?
2. Would it make sense to propose this feature for the Rust standard library, because Rust already supports `exec` on Unix systems? If it doesn't make sense to propose this for the Rust standard library, would it make sense to convert all the Unix code in this library to a thin wrapper over the standard library version?
3. On Unix, if process A runs process B, and process B execs process C, then the final exit code of process C will be returned to process A as if it were the exit code of process B. This is a key feature of `exec`. Does this work the same on Windows?

So basically, I have a bunch of "This is cool, but where do we actually go from here?" questions. If we do decide that this is the right way to proceed, then I also have a few minor comments on the implementation itself.

Once again, many thanks for tackling this! `exec` on Windows has been hard to do in Rust.

wfraser commented:
Thanks for the review! I'll address your main points:

> This is a fairly old library, and since Rust 1.9.0, `exec` has been [built into Rust](https://doc.rust-lang.org/std/os/unix/process/trait.CommandExt.html#tymethod.exec), but only for Unix-like systems. So this library had been largely deprecated. (I somehow missed updating the `Cargo.toml` file to indicate this.)

I was in the process of upgrading a program which used this crate, and discovered precisely what you point out, that the functionality in `std` is Unix-only, which means this crate has an opportunity to implement cross-platform functionality beyond what the standard library provides, making it useful again! :)

>     1. The [docs say that](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/execvp-wexecvp?view=vs-2019) "This API cannot be used in applications that execute in the Windows Runtime. For more information, see CRT functions not supported in Universal Windows Platform apps." Are normal Rust Windows apps supposed to work with the Windows Runtime? Are there any special `cfg` flags that we need to check or document?

Windows is one platform as far as Rust's `cfg` flags are concerned, even though Win32 and UWP are quite different. The main difference is that programs get linked with different libraries for the different environments. Win32 programs get the C runtime and other win32 DLLs, whereas UWP gets a whole different set of things.

The effect of trying to build this code on UWP will be to get a linker error when it can't find any library that implements the `_wexecvp` function.

>     2. Would it make sense to propose this feature for the Rust standard library, because Rust already supports `exec` on Unix systems? If it doesn't make sense to propose this for the Rust standard library, would it make sense to convert all the Unix code in this library to a thin wrapper over the standard library version?

Ideally yes, this functionality would be in the standard library, but adding things to the standard library is a really high bar that I don't think I'm up to meeting :P

Implementing the functionality in an external crate is often a desirable first step before proposing standard library changes anyway.
 
>     3. On Unix, if process A runs process B, and process B execs process C, then the final exit code of process C will be returned to process A as if it were the exit code of process B. This is a key feature of `exec`. Does this work the same on Windows?

Yes, the behavior is just as in Unix, including propagation of the exec'd process' exit code.

wfraser committed:
remove NulError from error type

while the NulError is nice in that it includes where the NUL is, it
prevents us from having the same API on Windows as on Unix, and
doesn't really add enough value to overcome that, so it's now
replaced with a nullary Error variant 'NullByteInArgument'.

Bumping version to 0.4.0 because this is a breaking change.

wfraser committed:
make _wexecvp definition top-level

wfraser committed:
examples/exec: don't pass the 0th arg twice

wfraser commented:
I've added `wexecvp` (and family) to `libc`. Once it is released, I can remove it from here and update the libc dependency.

emk commented:
This is great! Thank you for your responses. 

Obviously we're going to want to _remove_ the Unix implementation as well, and replace it with a call to Rust's standard library version.

gustavosimon commented:
Congratulations to developed this wrapper about the execvp function. I used it in a project, and it worked perfectly.
The use of the lib is very simple. 
But, I faced a limit that the lib just work to Linux OS. So I found this PR implementing the function to Windows OS.
Do you have any prevision to merge this PR and homologate the changes to use in production?

gustavosimon commented:
@emk have you a prevision to review this PR?

gustavosimon commented:
@wfraser @emk
